source/storing-data/provenance.rst

```diff
@@ -2,3 +2,213 @@ Decentralized Provenance Tracking
+
+   An added layer of complexity:
+   for Pipelines, the ``uuid`` in the execution block above is actually an alias UUID shared by all Pipeline Actions.
+   The ``alias-of`` UUID shown in the action block below describes the specific Action in that case.
+   This allows tools like ``q2view`` to nest all actions run by a single command within a single block.
+
+   For example:
+
+   The ``unweighted_unifrac_emperor.qza`` described below will have have three different IDs:
+
+   - The Result UUID, in ``metadata.yaml`` is unique to this Result
+   - The Action UUID, in ``action.yaml`` ``execution`` is unique to this Pipeline's current execution, and present in all pipeline Actions that occurred during this execution. (i.e. all Results from one run of ``core-metrics-phylogenetic`` share this ID)
+   - The ``alias-of`` UUID, in ``action.yaml`` ``action`` is unique to the specific Action, run by this Pipeline, which generated this Result
+
+   We chose to use `v4 UUIDs <https://docs.python.org/3/library/uuid.html>`_ for our unique IDs,
+   but there is nothing special about them that couldn't be handled by a different unique identifier scheme.
+   They're just IDs.
+
+.. _`action-block`:
+
+The action block
+~~~~~~~~~~~~~~~~
+Details about the action, including action and plugin names, inputs and parameters
+
+.. code-block:: YAML
+
+   action:
+      type: pipeline
+      plugin: !ref 'environment:plugins:diversity'
+      action: core_metrics_phylogenetic
+      inputs:
+      -   table: 34b07e56-27a5-4f03-ae57-ff427b50aaa1
+      -   phylogeny: a10d5d44-62c7-4322-afbe-c9811bcaa3e6
+      parameters:
+      -   sampling_depth: 1103
+      -   metadata: !metadata 'metadata.tsv'
+      -   n_jobs_or_threads: 1
+      output-name: unweighted_unifrac_emperor
+      alias-of: 2adb9f00-a692-411d-8dd3-a6d07fc80a01
+
+- The type field describes the *type of the Action*: a :term:`Method`, :term:`Visualizer`, or :term:`Pipeline`.
+- The plugin field describes the plugin which registered the Action, details about which can be found in ``action.yaml``'s ``environment:plugins`` section. ``!ref`` is a custom YAML tag defined `here <https://github.com/qiime2/qiime2/blob/6d8932eda130d4a9356f977fece2e252c135d0b9/qiime2/core/archive/provenance.py#L84>`_, Generally, these custom tags provide a way to express a structure not easily described by basic YAML.
+- Inputs lists the registered names of all :term:`inputs<Input>` to the Action, as well as the UUIDs of the passed inputs. Note the distinction between inputs and parameters.
+- Parameters lists registered parameter names, and the user-passed (or selected default) values.
+- ``output-name`` is the user-passed name of the file written by this Action.
+- ``alias-of``: an optional field, present if the Action was run as part of a QIIME 2 :term:`Pipeline`, representing the *actual* Action UUID rather than the Pipeline Alias. See maintainer note above for details.
```

Comment:
The `alias-of` is actually a result UUID, the action UUIDs are typically not used as references.

The way that pipelines handle provenance is by creating an "aliased" result for each output. So if you wanted to track within a pipeline, the algorithm is:

1. Identify the input UUIDs for the pipeline, and remember them.
2. Follow the alias-of into your /artifacts/ directory, this will be the last true output of any internal action.
3. If that output was an import, then some Python happened between your inputs and that import, and so there's not much of an internal workflow (the pipeline is closer to a method, where some ???, happened in the middle).
4. Otherwise, follow the provenance backwards until either 3 occurs, or you find the inputs from 1.

---

The system works this way to preserve reasonable "output" nodes so you can view the entire pipeline as a single action, while preserving the internal pipeline (e.g. once you follow the alias, the pipeline ceases to exist from your perspective, which is why you would want to keep track of the aliased-result's [the pointer, not the dereferenced entity] inputs, so that you know when the boundary of the actual pipelines was.)

Reply:
quick sketch of the above: [Notebook 5.pdf](https://github.com/qiime2/dev-docs/files/6850183/Notebook.5.pdf)

Reply:
looking at the sketch, the alias-of is actually probably the same direction as the inputs, so nothing is that weird from the graph's perspective, it's just there's an imaginary "aliasing" step, which combines the provenance of the pipeline with the provenance of the internal DAG.

Reply:
Thanks for the awesome writeup. I spent some time digging through .qzas last night, and put together a little summary for future me. LMK if any of this sounds faulty?

- All provenance nodes with 'alias-of' UUIDs are "fake" nodes, useful for generating views, but actually just duplicates of their last "real" "inner" action.
- "Terminal" pipeline outputs always alias a node of the same semantic type
- Different terminal outputs from the same pipeline will (generally) have different alias-of's, _because they are aliasing different inner nodes_. E.g an emperor plot aliases an emperor plot, while the PCoA results alias the inner PCoA results.
- Execution UUIDs are identical for all terminal outputs from a single pipeline execution. (All Viz, PCoA, and DistanceMatrix .qzxs from one core-metrics run will share the same Execution UUID in their root-level `action.yaml`)
- "Inner" ("nested") pipeline nodes are shared by terminal pipeline results. In the example below, dotted-outline boxes are "inner nodes", and solid-outline boxes are terminal/pipeline results. 
  - The terminal DistanceMatrix Z3 is the alias-of B9
  - The terminal PCoA result A5 is the alias-of Q2
  - Both have identical input UUIDs (indicating they're from the same pipeline run).
  - Given the above, and the order of the `core-metrics-phylogenetic` pipeline algorithm, B9 must be the parent node of Q2 in the provenance of both A5 and the viz A1.

![PipelineProvenance](https://user-images.githubusercontent.com/39198770/126429463-6bd50d28-400e-4994-9233-ae2a1d97ac3a.png)

Reply:
That is all correct! Re: your very last bullet point: I would just mention that "Inner" pipeline nodes MAY be shared by terminal pipeline results, but it would be perfectly reasonable to have a pipeline where the inner provenance is totally unrelated between two outputs, only because they are sourced from two unrelated inputs of the pipeline (or an import happened).

I suspect you already totally understand that, I just wanted to be more specific for future readers.

Reply:
Recognizing that this is a complex topic and not well covered by the text to date, I'm adding a Pipeline Provenance section at the end of this page to show how nested provenance works. 

It looks like I'll have about a week after I get back (before the 8/23 merge deadline) to respond to comments on that, but feel free to correct or drop that commit if that makes things easier. It's a standalone, so reverting shouldn't impact anything else here.